Replaces extools with auxtools; atmos overhaul

Putnam3145 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Extools has been replaced with auxtools, a Rust-based replacement.

LINDA has been mostly torn out and replaced with my own FEA follow-up, FDA, a highly parallel finite difference method atmospheric protocol which is far more consistent, fast and reasonable than the previous. It moves gas in a LINDA-like way, but never gets stuck, has heat properly moving around and has the same performance no matter how many pressure differences there are (yes, really). It doesn't break planetary atmos and is, in general, easier to understand.

## Why It's Good For The Game

Extools has been a bit of a disaster. It causes issues with planet atmos, it causes crashes when loading away missions, and is just... unsafe code. The move to Rust and designing of an atmos subsystem with it being written in a fast, safe, concurrent-able programming language to begin should bring a huge improvement.

## Changelog
:cl:
add: FDA, LINDA without the bookkeeping.
add: Putnamos, a simpler replacement for Monstermos. If I can get Monstermos to work, this will be Monstermos instead.
del: LINDA, the old active-turfs-based atmos subsystem.
del: Monstermos? I would rather not get rid of this, but I can't get it to work correctly.
server: Extools has been removed, and loading extools alongside auxtools will cause massive problems. If this is tested or merged, remove extools from the static files.
/:cl:

Putnam3145 committed:
broken but compiles and runs sans panics

Putnam3145 committed:
giasfclfubrehber

Putnam3145 committed:
auxtools atmos works now

Putnam3145 committed:
Made turf processing stop if overtime.

Putnam3145 committed:
Updated to latest attempts. Oy.

Putnam3145 committed:
More testing and such. It's super slow now!

Putnam3145 committed:
Superconductivity is gone. Only rustmos now.

Putnam3145 committed:
whoops more sanity checking

Putnam3145 committed:
compiling for cit's cpu

Putnam3145 committed:
A lot of testing and fiddling. Not ready for primetime.

Putnam3145 committed:
Always remember to commit, lord.

Putnam3145 committed:
Updated auxmos (see desc)

1. All callbacks are now auxmos-side, using new auxcallbacks.
2. Heat processing is now paused for and has been optimized by making it have fewer callbacks.

Putnam3145 committed:
Updated to latest auxmos ver, aoao

Putnam3145 committed:
Tweaks and such.

Putnam3145 committed:
lord help me

Putnam3145 committed:
mostly works!

occasional deadlock at last test; gonna keep checking that out.

Putnam3145 committed:
Added maptick.

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
forgot about funny linux

Putnam3145 committed:
thanks away missions, very cool

Putnam3145 committed:
Various fixes and performance tweaks

Putnam3145 committed:
linux gets help too

Putnam3145 committed:
Fixing runtimes; optimization.

Putnam3145 committed:
Toggle for heat; optimizations

Putnam3145 committed:
Fixed machinery not being processed.

Putnam3145 committed:
Huge optimizations+some fixes

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

MarinaGryphon commented on code/__DEFINES/subsystems.dm:
```diff
@@ -197,6 +198,19 @@
 #define RUNLEVELS_DEFAULT (RUNLEVEL_SETUP | RUNLEVEL_GAME | RUNLEVEL_POSTGAME)
 
+// SSair run section
+#define SSAIR_PIPENETS 1
```
Lines 245-254 need to be removed.

MarinaGryphon reviewed (changes requested):
code\__HELPERS\_extools_api.dm is not included.

MarinaGryphon commented on code/controllers/subsystem/time_track.dm:
```diff
@@ -16,33 +17,81 @@ SUBSYSTEM_DEF(time_track)
+			length(SSair.hotspots),
+			length(SSair.networks),
+			length(SSair.high_pressure_delta),
+			length(SSair.active_super_conductivity)
```
```suggestion
```

MarinaGryphon commented on code/controllers/subsystem/time_track.dm:
```diff
@@ -16,33 +17,81 @@ SUBSYSTEM_DEF(time_track)
+			SSair.cost_pipenets,
+			SSair.cost_rebuilds,
+			SSair.get_active_turfs(), //does not return a list, which is what we want
+			SSair.get_amt_excited_groups(),
```
```suggestion
```

MarinaGryphon commented on code/controllers/subsystem/time_track.dm:
```diff
@@ -16,33 +17,81 @@ SUBSYSTEM_DEF(time_track)
+			SSair.cost_superconductivity,
+			SSair.cost_pipenets,
+			SSair.cost_rebuilds,
+			SSair.get_active_turfs(), //does not return a list, which is what we want
```
```suggestion
```
These removed entries may need to be replaced with some of the new procs/vars, like `turf_process_time()`, `heat_process_time()`, `cost_post_process`, etc., depending on how logs are formatted.

MarinaGryphon reviewed (changes requested):
~~I found out a way to leave suggestions on unchanged lines too.~~

MarinaGryphon commented on code/game/world.dm:
```diff
@@ -7,46 +7,25 @@ GLOBAL_VAR_INIT(tgs_initialized, FALSE)
 	world.Profile(PROFILE_START)
+	AUXTOOLS_CHECK(AUXMAPTICK)
+	initialize_maptick()
+	log_world("World loaded at [TIME_STAMP("hh:mm:ss", FALSE)]!")
```
This is a duplicate of line 32.

Putnam3145 committed:
Makes it compile.

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
makes what i'm doing with model_temp more explicit

MarinaGryphon commented on code/__DEFINES/is_helpers.dm:
```diff
@@ -240,3 +240,5 @@ GLOBAL_LIST_INIT(glass_sheet_types, typecacheof(list(
 
 #define isProbablyWallMounted(O) (O.pixel_x > 20 || O.pixel_x < -20 || O.pixel_y > 20 || O.pixel_y < -20)
+
+#define isfinite(n) (isnum(n) && n == n)
```
This doesn't actually work.
```suggestion
#define isfinite(n) (x != 1.#IND && abs(x) < 1.#INF)
```
Handles NaN, positive infinity, and negative infinity.

MarinaGryphon reviewed (changes requested):
I made a mistake with the `isfinite` logic when testing it. It straight up doesn't work as-is.

Putnam3145 committed:
Removes explosive decomp; makes planet atmos check better

Putnam3145 committed:
Fixed some major incompetency from 2018.

// If someone will place 0 of some gas there, SHIT WILL BREAK. Do not do that.
LINDA_turf_tile.dm, line 16.

Putnam3145 committed:
Maybe fixed the shuttle badness

Putnam3145 committed:
Some extra comments

Putnam3145 committed:
Updates the .dlls for new server

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
Removed an unfortunate add_to_active

Putnam3145 committed:
Added a zone-based optimization.

Putnam3145 committed:
Revert "Added a zone-based optimization."

This reverts commit e1fbb27be5dfd37b746c9b7156d88e7bedfcd300.

Putnam3145 committed:
okay let me just try the normal thing then.

Putnam3145 committed:
fixed pipenets unless it's CPU-specific (oh no)

Putnam3145 committed:
Updated to 514.

Putnam3145 committed:
that should NOT be there

Putnam3145 committed:
ensured consistency with air-spawning atoms

Putnam3145 committed:
Makes adjust_moles clamp to valid values

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
i forgot i renamed this

Putnam3145 committed:
Fixed putnamos skipping space tiles

Putnam3145 committed:
Fixes initial turf temperatures

Putnam3145 committed:
Paranoid updates to maybe fix air duplication

Putnam3145 committed:
Fixes reactions breaking and maybe scrubber duping?

silicons committed:
Merge remote-tracking branch 'citadel/master' into auxtools-atmos

silicons committed:
missed that

Putnam3145 committed:
Well, that works

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
A minor thread optimization.

Putnam3145 committed:
I forgot to use the latest DLL, whoops.

Putnam3145 committed:
these are atmos machines. learn to program

Putnam3145 committed:
quick hack to fix corruption

Putnam3145 committed:
i made it work again

Putnam3145 committed:
works but leaks

Putnam3145 committed:
Various backend tweaks

Putnam3145 committed:
ACTUALLY LEAKS BUT WORKS WHOOPS

Putnam3145 committed:
and an extra check for stuff

Putnam3145 committed:
replaces Del with Destroy

Putnam3145 committed:
no this absolutely has to be Del.

Putnam3145 committed:
extreme optimization project to maybe fix game

Putnam3145 committed:
this actually does not need named this way

Putnam3145 committed:
Forced extra turf refcount tracking

Putnam3145 committed:
WHOOPS I SWAPPED FILTERS

Putnam3145 committed:
actually fixed the bug causing me grief

Putnam3145 committed:
minor optimization change

Putnam3145 committed:
added another cleanup i missed

what is wrong with me

Putnam3145 committed:
auxgm part 1

see also https://github.com/Baystation12/Baystation12/blob/dev/code/modules/xgm/xgm_gas_data.dm

Putnam3145 committed:
More aggressive garbage collection

Putnam3145 committed:
Fixed the funny UI stuff

Putnam3145 committed:
the linter is your friend

Putnam3145 committed:
Started expanding auxgm a bit

Putnam3145 committed:
Aaaa

Putnam3145 committed:
Finished the first auxgm expansion

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
Updates to master

Closes #14823

Putnam3145 committed:
i messed up mole adjusting

Putnam3145 committed:
fixes scrubbers oh no

Putnam3145 committed:
aoaoaoao

Putnam3145 committed:
makes o2 not just sorta disappear when breathed

Putnam3145 committed:
optimization, fix to breath, experimenting with speed

Putnam3145 committed:
Fixed deactivated supermatters eating gas

Putnam3145 committed:
Fixed the scrubber issue (oofies)

Putnam3145 committed:
Makes trit fires in pipenets not runtime

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
Fixes some scrubbing + garbage collect stuff

Putnam3145 committed:
Fixed a bug, accelerated gas mixture del

Putnam3145 committed:
Makes heat capacity just uncache on multiply

instead of trying weird tricks

Putnam3145 committed:
Made hotspots take generic fires into account

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into auxtools-atmos

Putnam3145 committed:
Jank fix

Putnam3145 committed:
Makes closed turfs get removed from atmos adjacent.

Putnam3145 committed:
fix double . in fusion scan

Putnam3145 committed:
actually makes the default gas damage work, whoops

Putnam3145 committed:
okay, we DO have to check if it's not null

Putnam3145 committed:
Fixes the varedit interface

Putnam3145 committed:
Some baffling optimizations

Putnam3145 committed:
minor optimizations

Putnam3145 committed:
Wow It's Been At Double Speed

Putnam3145 committed:
Minor backend cleanup
